content/blog/functional-programming-constructs/index.md

```diff
@@ -0,0 +1,155 @@
+---
+title: 'Functional programming constructs'
+description: ''
+date: 2019-10-11
+categories: ['javascript', 'functional-programming']
+ogImage: ./og-image.png
+---
+
+[Functional programming](https://en.wikipedia.org/wiki/Functional_programming) is a **declarative programming paradigm** that focuses on **composition** using **pure functions**. It **avoids mutation** of state and side-effects (instead encouraging us to **isolate side effects** and make them more predictable in the code).
+
+I've recently started to dive deeper into functional programming and was quite elated to know functional programming has its roots in [lambda calculus](https://en.wikipedia.org/wiki/Lambda_calculus). A lot of techniques and terms in functional programming are derived from mathematics _(perhaps I should have paid more attention to engineering mathematics at college)_.
+
+## Why functional programming?
+
+Functional programming aims at making code more readable and understandable. We spend more time reading code than writing it, which justifies the returns to me.
+
+> Code that you cannot trust is code that you do not understand
+>
+> — [Kyle Simpson](https://twitter.com/getify)
+
+Consider an expression using the `map` keyword. Same logic can be written using a `for` loop, however with the `for` loop we've to read through the entire block to figure out what it does, condition for breaking out of the loop and the values being changed (mutated) inside the `for` loop. When using a `map` a lot of this is implicitly understood by the programmer, therefore less code to read. For example:
+
+```js
+// Create a new array from an existing array incrementing each value by 1
+// [1, 2, 3] -> [2, 3, 4]
+
+const originalArray = [1, 2, 3];
+
+// First approach using a loop
+let newArray = [];
+// Note: don't use for...in for iterating over arrays since it returns array indices unordered
+for (let value of originalArray) {
+	newArray.push(value + 1);
```

Comment:
fix identation

Reply:
The file runs through prettier, so indentation is auto formatted.